### PR TITLE
Add configuration variable expansion

### PR DESCRIPTION
**Pull request by jhakonen**
This PR implements issue #560.

I went with the variable pattern of `${ENV:NAME}` or `$ENV:NAME` as suggested by jpmens. Including also support for loading file content with `${FILE:file/path}`. I used FILE instead of FS as I think it is more explicit.

I didn't add support for HashiCorp's Vault as I don't have personal experience of using it, but provided that there's a way to fetch vault's value then it is easy to add support for it by modifying `VariableInterpolation.sources`.

### CODE REVIEW

**Review by jpmens — APPROVED** [review]
Looks good to me, and thank you for your contribution!

**Comment by amotl** [comment]
Thank you very much, Janne! 💯

File: tests/test_configuration.py [hunk 1]
```diff
@@ -81,3 +83,85 @@ def test_config_better_addresses_pushsafer():
...
+                "PÄSSWÖRD_3": "non-ascii-secret",
+                "/path/to/password.txt": "file-contents",
+            }
+        ),
+        "SRC_2": create_source(
+            {
+                "PASSWORD_1": "p4ssw0rd",
+            }
+        ),
+    }
```
Comment (amotl) [inline tests/test_configuration.py:143]: `SRC_1` and `SRC_2` are two synthetic variable interpolation types like `ENV` or `FILE`, but only used within the test suite?

**Review by amotl — CHANGES_REQUESTED** [review]
Dear Janne,

thanks for your excellent patch. I've just added a question about the test suite.

With kind regards,
Andreas.

File: tests/test_configuration.py (hunk 1 above)
Comment (jhakonen) [reply tests/test_configuration.py:143]: Yes. I'm testing `expand_vars()` function here directly, bypassing concrete implementation of `ENV` and `FILE` sources, so I've replaced them with these fakes.

File: tests/test_configuration.py [hunk 2]
```diff
@@ -81,3 +83,85 @@ def test_config_better_addresses_pushsafer():
...
+def test_expand_vars_variable_not_found():
+    """
+    Verify that `expand_vars` raises error when variable is not in source.
+    """
+
+    def empty_source(name):
+        raise KeyError("Variable not found")
+
+    with pytest.raises(KeyError, match=re.escape("$SRC_1:VARIABLE: 'Variable not found'")):
+        mqttwarn.configuration.expand_vars("-->$SRC_1:VARIABLE<--", {"SRC_1": empty_source})
```
[the remark this answers is not in the document]
Comment (jhakonen) [reply tests/test_configuration.py:167]: These test functions shouldn't affect each other as there is no global state used in `expand_vars()`. It doesn't have any side-effect when called.

Here in this test function we're passing in a new dict of sources `{"SRC_1": empty_source}` which is not same dict as the `sources` in `test_expand_vars_ok()` test function.

I tried moving `test_expand_vars_variable_not_found()` before `test_expand_vars_ok()` but it passes still fine:
```bash
[nix-shell:~/mqttwarn]$ pytest --no-header --disable-warnings --no-cov -k test_expand_vars
=================================================== test session starts ===================================================
collected 318 items / 304 deselected / 1 skipped / 14 selected                                                            

tests/test_configuration.py::test_expand_vars_variable_not_found PASSED                                             [  7%]
tests/test_configuration.py::test_expand_vars_ok[my-password-my-password] PASSED                                    [ 14%]
tests/test_configuration.py::test_expand_vars_ok[$SRC_1:PASSWORD_1-my-password] PASSED                              [ 21%]
tests/test_configuration.py::test_expand_vars_ok[$SRC_1:PASSWORD_2-super-secret] PASSED                             [ 28%]
tests/test_configuration.py::test_expand_vars_ok[-->$SRC_1:PASSWORD_1<----->my-password<--] PASSED                  [ 35%]
tests/test_configuration.py::test_expand_vars_ok[$SRC_2:PASSWORD_1-p4ssw0rd] PASSED                                 [ 42%]
tests/test_configuration.py::test_expand_vars_ok[$SRC_1:P\xc4SSW\xd6RD_3-non-ascii-secret] PASSED                   [ 50%]
tests/test_configuration.py::test_expand_vars_ok[${SRC_1:PASSWORD_1}-my-password] PASSED                            [ 57%]
tests/test_configuration.py::test_expand_vars_ok[${SRC_1:/path/to/password.txt}-file-contents] PASSED               [ 64%]
tests/test_configuration.py::test_expand_vars_ok[${SRC_1:PASSWORD_1} ${SRC_1:PASSWORD_2}-my-password super-secret] PASSED [ 71%]
tests/test_configuration.py::test_expand_vars_ok[$SRC_1:PASSWORD_1 ${SRC_1:/path/to/password.txt} $SRC_1:PASSWORD_1-my-password file-contents my-password] PASSED [ 78%]
tests/test_configuration.py::test_expand_vars_ok[${SRC_1:/path/to/password.txt} $SRC_1:PASSWORD_1 ${SRC_1:/path/to/password.txt}-file-contents my-password file-contents] PASSED [ 85%]
tests/test_configuration.py::test_expand_vars_ok[/$SRC_1:PASSWORD_1/$SRC_1:PASSWORD_2/foo.txt-/my-password/super-secret/foo.txt] PASSED [ 92%]
tests/test_configuration.py::test_expand_vars_variable_type_not_supported PASSED                                    [100%]

================================================= short test summary info =================================================
SKIPPED [1] tests/services/test_apns.py:11: The `apns` package is not ready for Python3
================================ 14 passed, 1 skipped, 304 deselected, 1 warning in 1.06s =================================
```

[the remark this answers is not in the document]
Comment (amotl) [reply tests/test_configuration.py:167]: I see, probably missed to spot the point. Thanks for clarifying!

**Pull request merged**